File: linear_search/linear.py
import random
import time
import csv

REPEATS = 5000

def linear_search(arr, key):
    for i in range(len(arr)):
        if arr[i] == key:
            return i
    return -1

with open("linear_average.csv", "w", newline="") as file:
    writer = csv.writer(file)
    writer.writerow(["Array_Size", "Average_Time_sec"])

    random.seed(123)

    for n in range(10000, 100001, 5000):
        arr = [random.randint(1, 10**9) for _ in range(n)]

        total_time = 0

        for _ in range(REPEATS):
            key = random.randint(1, 10**9)

            start = time.perf_counter_ns()
            linear_search(arr, key)
            end = time.perf_counter_ns()

            total_time += (end - start)

        avg_time = total_time / (REPEATS * 1e9)
        writer.writerow([n, f"{avg_time:.12f}"])
        print(f"Linear Search done for n = {n}")

File: linear_search/linear.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <random>
#include <chrono>

using namespace std;
using namespace std::chrono;

int linearSearch(const vector<int>& arr, int key)
{
    for (int i = 0; i < arr.size(); i++)
        if (arr[i] == key)
            return i;
    return -1;
}

int main()
{
    ofstream fout("linear_average.csv");
    fout << "Array_Size,Average_Time_ns\n";

    mt19937 rng(123);
    uniform_int_distribution<int> dist;

    const int REPEATS = 10000;

    for (int n = 10000; n <= 100000; n += 5000)
    {
        vector<int> arr(n);
        for (int i = 0; i < n; i++)
            arr[i] = rng();

        dist = uniform_int_distribution<int>(0, INT32_MAX);

        long long totalTime = 0;

        for (int i = 0; i < REPEATS; i++)
        {
            int key = dist(rng);

            auto start = high_resolution_clock::now();
            linearSearch(arr, key);
            auto end = high_resolution_clock::now();

            totalTime += duration_cast<nanoseconds>(end - start).count();
        } 

        double avgTime = (double)totalTime / REPEATS;
        fout << n << "," << avgTime << "\n";

        cout << "Linear Search done for n = " << n << endl;
    }

    fout.close();
    return 0;
}